--- templates/create-mailbox.php	2007-12-02 14:23:08.000000000 -0500
+++ templates/create-mailbox.php	2008-07-04 09:59:56.000000000 -0400
@@ -45,7 +45,8 @@
    <?php if ($CONF['quota'] == 'YES') { ?>
    <tr>
       <td><?php print $PALANG['pCreate_mailbox_quota'] . ":"; ?></td>
-      <td><input class="flat" type="text" name="fQuota" value="<?php print $tQuota; ?>" /></td>
+      <!--<td><input class="flat" type="text" name="fQuota" value="<?php print $tQuota; ?>" /></td>-->
+      <td><input class="flat" type="text" name="fQuota" value="10" /></td>
       <td><?php print $pCreate_mailbox_quota_text; ?></td>
    </tr>
    <?php } ?>
@@ -56,7 +57,8 @@
    </tr>
    <tr>
       <td><?php print $PALANG['pCreate_mailbox_mail'] . ":"; ?></td>
-      <td><input class="flat" type="checkbox" name="fMail" <?php print (isset($CONF['create_mailbox_subdirs'])) ? '' : 'checked'; ?> /></td>
+      <!--<td><input class="flat" type="checkbox" name="fMail" <?php print (isset($CONF['create_mailbox_subdirs'])) ? '' : 'checked'; ?> /></td>-->
+      <td><input class="flat" type="checkbox" name="fMail" /></td>
       <td>&nbsp;</td>
    </tr>
    <tr>
