diff -Naur --exclude='*.svn*' roundcube_svn/plugins/changepasswd/changepasswd.php roundcube_pub/plugins/changepasswd/changepasswd.php
--- roundcube_svn/plugins/changepasswd/changepasswd.php	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/plugins/changepasswd/changepasswd.php	2009-03-10 18:49:10.000000000 +0000
@@ -0,0 +1,31 @@
+<?php
+
+/*
+ +-----------------------------------------------------------------------+
+ | plugins/changepasswd/chnagepasswd.php                                 |
+ |                                                                       |
+ | This file is part of the RoundCube Webmail client                     |
+ | Copyright (C) 2008, RoundCube Dev. - Switzerland                      |
+ | Licensed under the GNU GPL                                            |
+ |                                                                       |
+ | PURPOSE:                                                              |
+ |   Provide functionality for user's password modification via sql      |
+ |                                                                       |
+ +-----------------------------------------------------------------------+
+ | Author: Philip Weir                                                   |
+ +-----------------------------------------------------------------------+
+
+ $Id: $
+
+*/
+
+class changepasswd extends rcube_plugin
+{
+	function init()
+	{
+		// preperation for Plugin API
+		// not used yet
+	}
+}
+
+?>
\ No newline at end of file
diff -Naur --exclude='*.svn*' roundcube_svn/plugins/changepasswd/config.inc.php roundcube_pub/plugins/changepasswd/config.inc.php
--- roundcube_svn/plugins/changepasswd/config.inc.php	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/plugins/changepasswd/config.inc.php	2009-03-10 18:49:10.000000000 +0000
@@ -0,0 +1,42 @@
+<?php
+
+/*
+ +-----------------------------------------------------------------------+
+ | Changepasswd configuration file                                       |
+ |                                                                       |
+ | This file is part of the RoundCube Webmail client                     |
+ | Copyright (C) 2005-2008, RoundCube Dev. - Switzerland                 |
+ | Licensed under the GNU GPL                                            |
+ |                                                                       |
+ +-----------------------------------------------------------------------+
+
+*/
+
+$changepasswd_config = array();
+
+// User database settings
+$changepasswd_config['db_dsnw'] = 'mysql://username:password@localhost/database';
+
+// PEAR database DSN for read only operations (if empty write database will be used)
+// useful for database replication
+$changepasswd_config['db_dsnr'] = '';
+
+// use persistent db-connections
+// beware this will not "always" work as expected
+// see: http://www.php.net/manual/en/features.persistent-connections.php
+$changepasswd_config['db_persistent'] = FALSE;
+
+// Database name and table
+$changepasswd_config['db_name_table'] = 'vmail.mailbox';
+
+// Username filed
+$changepasswd_config['db_user_field'] = 'username';
+
+// Password filed
+$changepasswd_config['db_passwd_field'] = 'password';
+
+?>
\ No newline at end of file
diff -Naur --exclude='*.svn*' roundcube_svn/program/js/changepasswd.js roundcube_pub/program/js/changepasswd.js
--- roundcube_svn/program/js/changepasswd.js	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/program/js/changepasswd.js	2009-03-10 18:49:10.000000000 +0000
@@ -0,0 +1,43 @@
+/*
+ +-----------------------------------------------------------------------+
+ | RoundCube Webmail Client Script                                       |
+ |                                                                       |
+ | This file is part of the RoundCube Webmail client                     |
+ | Copyright (C) 2008, RoundCube Dev, - Switzerland                      |
+ | Licensed under the GNU GPL                                            |
+ |                                                                       |
+ +-----------------------------------------------------------------------+
+ | Author: Philip Weir                                                   |
+ +-----------------------------------------------------------------------+
+
+ $Id: $
+
+*/
+
+function changepasswd_check_input() {
+	var input_curpasswd = rcube_find_object('_curpasswd');
+	var input_newpasswd = rcube_find_object('_newpasswd');
+	var input_confpasswd = rcube_find_object('_confpasswd');
+	
+	if (input_curpasswd && input_curpasswd.value == '') {
+		alert(rcmail.get_label('nocurpassword'));
+		input_curpasswd.focus();
+	}        
+	else if (input_newpasswd && input_newpasswd.value == '' && input_confpasswd && input_confpasswd.value == '') {
+		alert(rcmail.get_label('nopassword'));
+		input_newpasswd.value = '';
+		input_confpasswd.value = '';
+		input_newpasswd.focus();
+	}
+	else if (input_newpasswd && input_confpasswd && input_newpasswd.value != input_confpasswd.value) {
+		alert(rcmail.get_label('passwordinconsistency'));
+		input_newpasswd.value = '';
+		input_confpasswd.value = '';
+		input_newpasswd.focus();
+	}
+	else {
+		return true;
+	}
+	
+	return false;
+}
diff -Naur --exclude='*.svn*' roundcube_svn/program/js/app.js roundcube_pub/program/js/app.js
--- roundcube_svn/program/js/app.js	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/program/js/app.js	2009-03-10 18:49:10.000000000 +0000
@@ -289,6 +289,7 @@
       case 'settings':
         this.enable_command('preferences', 'identities', 'save', 'folders', true);
         this.enable_command('managesieve', true);
+        this.enable_command('changepasswd', true);
         
         if (this.env.action=='identities' || this.env.action=='edit-identity' || this.env.action=='add-identity') {
           this.enable_command('add', this.env.identities_level < 2);
@@ -318,6 +319,9 @@
              this.enable_command('sievefiltersave', true);
          }
    
+        if (this.env.action=='changepasswd')
+          this.enable_command('save-changepasswd', true);
+
         if (this.gui_objects.identitieslist)
           {
           this.identity_list = new rcube_list_widget(this.gui_objects.identitieslist, {multiselect:false, draggable:false, keyboard:false});
@@ -1053,6 +1057,15 @@
         this.goto_url('identities');
         break;
           
+      case 'changepasswd':
+        this.goto_url('changepasswd');
+        break;
+
+      case 'save-changepasswd':
+        if (changepasswd_check_input())
+          this.gui_objects.editform.submit();
+        break;
+
       case 'delete-identity':
         this.delete_identity();
         
diff -Naur --exclude='*.svn*' roundcube_svn/program/localization/en_US/labels.inc roundcube_pub/program/localization/en_US/labels.inc
--- roundcube_svn/program/localization/en_US/labels.inc	2009-01-18 11:24:41.000000000 +0000
+++ roundcube_pub/program/localization/en_US/labels.inc	2009-03-10 18:49:10.000000000 +0000
@@ -323,6 +323,11 @@
 $labels['MB'] = 'MB';
 $labels['GB'] = 'GB';
 
+$labels['changepasswd']  = 'Change password';
+$labels['curpasswd']  = 'Current password';
+$labels['newpasswd']  = 'New password';
+$labels['confpasswd']  = 'Retype new password';
+
 $labels['filters'] = 'Filters';
 $labels['managefilters'] = 'Manage incoming mail filters';
 $labels['filtername'] = 'Filter name';
diff -Naur --exclude='*.svn*' roundcube_svn/program/localization/en_US/messages.inc roundcube_pub/program/localization/en_US/messages.inc
--- roundcube_svn/program/localization/en_US/messages.inc	2009-01-18 11:24:41.000000000 +0000
+++ roundcube_pub/program/localization/en_US/messages.inc	2009-03-10 18:49:10.000000000 +0000
@@ -95,6 +95,14 @@
 $messages['nofromaddress'] = 'Missing e-mail address in selected identity';
 $messages['editorwarning'] = 'Switching to the plain text editor will cause all text formatting to be lost. Do you wish to continue?';
 
+$messages['nocurpassword'] = 'Please enter your current password';
+$messages['nopassword'] = 'Please enter your new password';
+$messages['passwordinconsistency'] = 'Inconsistency in new password, please try again';
+$messages['passwordconnerror'] = 'Error: Cannot connect to server';
+$messages['passwordsrverror'] = 'Error: Unknown server error';
+$messages['passwordfailed'] = 'Error: Current password incorrect';
+$messages['passwordchanged'] = 'Successfully changed password';
+
 $messages['filterunknownerror'] = 'Unknown server error';
 $messages['filterconnerror'] = 'Unable to connect to managesieve server';
 $messages['filterdeleteerror'] = 'Unable to delete filter. Server error occured';
diff -Naur --exclude='*.svn*' roundcube_svn/program/steps/settings/changepasswd.inc roundcube_pub/program/steps/settings/changepasswd.inc
--- roundcube_svn/program/steps/settings/changepasswd.inc	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/program/steps/settings/changepasswd.inc	2009-03-10 18:49:10.000000000 +0000
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ +-----------------------------------------------------------------------+
+ | plugins/changepasswd/changepasswd.inc                                 |
+ |                                                                       |
+ | This file is part of the RoundCube Webmail client                     |
+ | Copyright (C) 2008, RoundCube Dev. - Switzerland                      |
+ | Licensed under the GNU GPL                                            |
+ |                                                                       |
+ | PURPOSE:                                                              |
+ |   Provide functionality for user password modification                |
+ |                                                                       |
+ +-----------------------------------------------------------------------+
+ | Author: Philip Weir                                                   |
+ +-----------------------------------------------------------------------+
+
+ $Id: $
+
+*/
+
+$OUTPUT->set_pagetitle(rcube_label('changepasswd'));
+$OUTPUT->add_handler('changepasswd', 'rcmail_changepasswd_form');
+$OUTPUT->include_script('changepasswd.js');
+$OUTPUT->add_label('nocurpassword', 'nopassword', 'passwordinconsistency');
+$OUTPUT->send('changepasswd');
+
+function rcmail_changepasswd_form($attrib) {
+	list($form_start, $form_end) = get_form_tags($attrib, 'save-changepasswd');
+	unset($attrib['form']);
+	
+	// return the complete edit form as table
+	$out = $form_start;
+	$table = new html_table(array('cols' => 2));
+
+	// show old password field
+	if (!isset($no_override['curpasswd']))
+	{
+		$field_id = 'rcmfd_curpwd';
+		$input_curpasswd = new html_passwordfield(array('name' => '_curpasswd', 'id' => $field_id));
+		
+		$table->add('title', html::label($field_id, Q(rcube_label('curpasswd'))));
+		$table->add(null, $input_curpasswd->show());
+	}
+	  
+	// show new password field
+	if (!isset($no_override['newpasswd']))
+	{
+		$field_id = 'rcmfd_newpwd';
+		$input_newpasswd = new html_passwordfield(array('name' => '_newpasswd', 'id' => $field_id,));
+		
+		$table->add('title', html::label($field_id, Q(rcube_label('newpasswd'))));
+		$table->add(null, $input_newpasswd->show());
+	}
+	
+	// show new password confirm field
+	if (!isset($no_override['confpasswd']))
+	{
+		$field_id = 'rcmfd_cnfpwd';
+		$input_confpasswd = new html_passwordfield(array('name' => '_confpasswd', 'id' => $field_id));
+		
+		$table->add('title', html::label($field_id, Q(rcube_label('confpasswd'))));
+		$table->add(null, $input_confpasswd->show());
+	}
+	
+	$out .= $table->show($attrib);
+	$out .= $form_end;
+	
+	return $out;  
+}
+
+?>
\ No newline at end of file
diff -Naur --exclude='*.svn*' roundcube_svn/program/steps/settings/save_changepasswd.inc roundcube_pub/program/steps/settings/save_changepasswd.inc
--- roundcube_svn/program/steps/settings/save_changepasswd.inc	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/program/steps/settings/save_changepasswd.inc	2009-03-10 18:49:10.000000000 +0000
@@ -0,0 +1,69 @@
+<?php
+
+/*
+ +-----------------------------------------------------------------------+
+ | plugins/changepasswd/save_changepasswd.inc                            |
+ |                                                                       |
+ | This file is part of the RoundCube Webmail client                     |
+ | Copyright (C) 2008, RoundCube Dev. - Switzerland                      |
+ | Licensed under the GNU GPL                                            |
+ |                                                                       |
+ | PURPOSE:                                                              |
+ |   Save user account password                                          |
+ |                                                                       |
+ +-----------------------------------------------------------------------+
+ | Author: Philip Weir                                                   |
+ +-----------------------------------------------------------------------+
+ | Requires: plugins/changepasswd/config.inc.php                         |
+ +-----------------------------------------------------------------------+
+
+ $Id: $
+
+*/
+
+require_once('plugins/changepasswd/config.inc.php');
+
+$username = $_SESSION['username'];
+$old_pw = $_POST['_curpasswd'];
+$new_pw = $_POST['_newpasswd'];
+
+$user_db = new rcube_mdb2($changepasswd_config['db_dsnw'], $changepasswd_config['db_dsnr'], $changepasswd_config['db_persistent']);
+$user_db->db_connect('w');
+
+// check DB connections and exit on failure
+if ($err_str = $user_db->is_error()) {
+  raise_error(array(
+    'code' => 603,
+    'type' => 'db',
+    'message' => $err_str), FALSE, TRUE);
+}
+if ($_SESSION['password'] == $RCMAIL->encrypt_passwd($old_pw)) {
+    $updated = false;
+    //$updated = $user_db->query($sql_update);
+    $tmpEncPass = crypt($new_pw, '');
+    $updated = $user_db->query(
+        "UPDATE " . $changepasswd_config['db_name_table'] . 
+        " SET " . $changepasswd_config['db_passwd_field'] . " = '" . $tmpEncPass . "'" . 
+        " WHERE " . $changepasswd_config['db_user_field'] . " = '" . $username . "'"
+    );
+
+    if ($updated) {
+        $_SESSION['password'] = $tmpEncPass;
+        $OUTPUT->show_message('passwordchanged', 'confirmation');
+    }
+    else {
+        $OUTPUT->show_message('passwordfailed', 'error');
+    }
+}
+else {
+  	$OUTPUT->show_message('passwordfailed', 'error');
+}
+
+// go to next step
+rcmail_overwrite_action('changepasswd');
+
+?>
\ No newline at end of file
diff -Naur --exclude='*.svn*' roundcube_svn/skins/default/includes/settingstabs.html roundcube_pub/skins/default/includes/settingstabs.html
--- roundcube_svn/skins/default/includes/settingstabs.html	2008-05-20 17:45:39.000000000 +0100
+++ roundcube_pub/skins/default/includes/settingstabs.html	2009-03-10 18:49:10.000000000 +0000
@@ -2,5 +2,6 @@
 <span id="settingstabdefault" class="tablink"><roundcube:button command="preferences" type="link" label="preferences" title="editpreferences" /></span>
 <span id="settingstabfolders" class="tablink"><roundcube:button command="folders" type="link" label="folders" title="managefolders" class="tablink" /></span>
 <span id="settingstabidentities" class="tablink"><roundcube:button command="identities" type="link" label="identities" title="manageidentities" class="tablink" /></span>
+<span id="settingstabchangepasswd" class="tablink"><roundcube:button command="changepasswd" type="link" label="password" title="changepasswd" class="tablink" /></span>
 <span id="settingstabmanagesieve" class="tablink"><roundcube:button command="managesieve" type="link" label="filters" title="managefilters" class="tablink" /></span>
 </div>
diff -Naur --exclude='*.svn*' roundcube_svn/skins/default/templates/changepasswd.html roundcube_pub/skins/default/templates/changepasswd.html
--- roundcube_svn/skins/default/templates/changepasswd.html	1970-01-01 01:00:00.000000000 +0100
+++ roundcube_pub/skins/default/templates/changepasswd.html	2009-03-10 18:49:10.000000000 +0000
@@ -0,0 +1,27 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml">
+<head>
+<title><roundcube:object name="pagetitle" /></title>
+<roundcube:include file="/includes/links.html" />
+<link rel="stylesheet" type="text/css" href="/settings.css" />
+<script type="text/javascript" src="/functions.js"></script>
+</head>
+<body onload="rcube_init_settings_tabs()">
+
+<roundcube:include file="/includes/taskbar.html" />
+<roundcube:include file="/includes/header.html" />
+<roundcube:include file="/includes/settingstabs.html" />
+
+<div id="userprefs-box">
+<div id="userprefs-title"><roundcube:label name="changepasswd" /></div>
+
+<div style="padding:15px 0 15px 15px">
+<roundcube:object name="changepasswd" />
+<div style="clear:left"></div>			
+</div>
+</div>
+
+<p id="listbuttons"><roundcube:button command="save-changepasswd" type="input" class="button mainaction" label="save" /></p>
+		
+</body>
+</html>
