--- program/steps/mail/get.inc.bak	2008-12-17 13:27:12.000000000 +0800
+++ program/steps/mail/get.inc	2008-12-17 15:31:35.000000000 +0800
@@ -106,9 +106,19 @@
       }
     else
       {
-      header(sprintf('Content-Disposition: %s; filename="%s";',
+      $HTTP_USER_AGENT = $_SERVER['HTTP_USER_AGENT'];
+
+      if (strstr($HTTP_USER_AGENT, 'compatible; MSIE ') ||
+          strstr($HTTP_USER_AGENT, 'Chrome')) {
+          header(sprintf('Content-Disposition: %s; filename="%s";',
+                    $_GET['_download'] ? 'attachment' : 'inline',
+                    $part->filename ? rawurlencode(abbreviate_string($part->filename, 55)) :
+                                      rawurlencode("roundcube.$ctype_secondary")));
+      } else {
+          header(sprintf('Content-Disposition: %s; filename="%s";',
                      $_GET['_download'] ? 'attachment' : 'inline',
                      $part->filename ? abbreviate_string($part->filename, 55) : "roundcube.$ctype_secondary"));
+      }
 
       // turn off output buffering and print part content
       $IMAP->get_message_part($MESSAGE['UID'], $part->mime_id, $part, true);
