--- roundcube_svn/program/localization/zh_CN/labels.inc	2009-01-17 09:55:39.000000000 +0800
+++ roundcube_pub/program/localization/zh_CN/labels.inc	2009-04-13 15:44:05.000000000 +0800
@@ -265,4 +265,74 @@
 $labels['MB'] = 'MB';
 $labels['GB'] = 'GB';
 
+$labels['filters'] = '过滤器';
+$labels['managefilters'] = '管理邮件过滤器';
+$labels['filtername'] = '过滤器名称';
+$labels['disablerule'] = '禁用这条过滤规则';
+$labels['disabled'] = '已禁用';
+$labels['newfilter'] = '新建过滤器';
+$labels['moveup'] = '上移';
+$labels['movedown'] = '下移';
+$labels['filterallof'] = '匹配下面的所有规则';
+$labels['filteranyof'] = '匹配下面的任意一条规则';
+$labels['filterany'] = '所有邮件';
+$labels['filtercontains'] = '包含';
+$labels['filternotcontains'] = '不包含';
+$labels['filteris'] = '等于';
+$labels['filterisnot'] = '不等于';
+$labels['filterexists'] = '存在';
+$labels['filternotexists'] = '不存在';
+$labels['filterregex'] = '匹配正则表达式';
+$labels['filternotregex'] = ' 不匹配正则表达式';
+$labels['filterunder'] = '小于';
+$labels['filterover'] = '大于';
+$labels['filteradvoptions'] = '更多选项...';
+$labels['operator'] = '操作';
+$labels['comparator'] = '比较';
+$labels['count'] = '数量';
+$labels['value'] = '值';
+$labels['isgreaterthan'] = '大于';
+$labels['isgreaterthanequal'] = '大于或等于';
+$labels['islessthan'] = '小于';
+$labels['islessthanequal'] = '小于或等于';
+$labels['equals'] = '等于';
+$labels['notequals'] = '不等于';
+$labels['userpart'] = '用户名字符串等于';
+$labels['notuserpart'] = '用户名字符串不等于';
+$labels['detailpart'] = 'detail part equals';
+$labels['notdetailpart'] = 'detail part does not equal';
+$labels['domainpart'] = '域名等于';
+$labels['notdomainpart'] = '域名不等于';
+$labels['teststring'] = '要检测的字符串';
+$labels['messagemoveto'] = '移动邮件到';
+$labels['messageredirect'] = '转发邮件到';
+$labels['messageimapflags'] = '将邮件标记为';
+$labels['messagereject'] = '拒收并回复消息';
+$labels['messagevacation'] = '外出自动回复消息';
+$labels['messagekeep'] = '保留邮件';
+$labels['messagediscard'] = '丢弃邮件';
+$labels['messagenotify'] = '发送通知';
+$labels['method'] = '方法/途径';
+$labels['messagesrules'] = '过滤规则';
+$labels['messagesactions'] = '过滤操作';
+$labels['flagread'] = '已读';
+$labels['flagdeleted'] = '已删除';
+$labels['flaganswered'] = '已回复';
+$labels['flagdraft'] = '草稿';
+$labels['flagflagged'] = '已标记';
+$labels['addsieverule'] = '在当前规则下添加新的规则';
+$labels['addsieveact'] = '当前操作下添加新的操作';
+$labels['deletesieverule'] = '删除这条规则';
+$labels['deletesieveact'] = '删除这个操作';
+$labels['envelopefrom'] = 'Envelope From';
+$labels['envelopeto'] = 'Envelope To';
+$labels['otherheader'] = '其它邮件头';
+$labels['days'] = '天（日期）';
+$labels['message'] = '信息';
+$labels['sieveruleheaders'] = '查看邮件头示例';
+
+$labels['changepasswd']  = '修改密码';
+$labels['curpasswd']  = '当前密码';
+$labels['newpasswd']  = '新密码';
+$labels['confpasswd']  = '重新输入新密码';
 ?>
--- roundcube_svn/program/localization/zh_CN/messages.inc	2009-01-17 09:55:39.000000000 +0800
+++ roundcube_pub/program/localization/zh_CN/messages.inc	2009-04-13 15:45:09.000000000 +0800
@@ -96,4 +96,44 @@
 $messages['nofromaddress'] = '选种的身份中没有邮件地址';
 $messages['editorwarning'] = '切换到纯文本编辑器将导致邮件正文中的所有文本格式失效，您确定要这样做吗？';
 
+$messages['nofromaddress'] = '选中的身份中没有邮件地址';
+$messages['nosieverules'] = '没有过滤规则。';
+$messages['filterdeleteconfirm'] = '您确定要删除所选择的过滤器吗？';
+$messages['ruledeleteconfirm'] = '您确定要删除所选择的规则吗？';
+$messages['actiondeleteconfirm'] = '您确定要删除所选择的过滤操作吗？';
+$messages['filterunknownerror'] = '未知的服务器错误';
+$messages['filterconnerror'] = '无法连接到 managesieve 服务器';
+$messages['filterdeleteerror'] = '无法删除过滤器。服务器错误';
+$messages['filterdeleted'] = '过滤规则已成功删除';
+$messages['filtersaved'] = '过滤规则已成功保存。';
+$messages['filtersaveerror'] = '无法保存过滤器。服务器错误';
+$messages['vacdaysexp'] = '间隔的天数，表示在这段时间内，不管对方给你发送多少次邮件，系统都只发送一条自动回复消息给他。';
+$messages['norulename'] = '请为这条过滤规则输入一个名称';
+$messages['ruleexists'] = '同名的过滤规则已经存在，请输入另外不同的名称';
+$messages['noheader'] = '请输入用于检测的邮件头名称';
+$messages['headerbadchars'] = '错误：邮件头包含非法字符';
+$messages['noheadervalue'] = '请输入一个用于比对邮件头的值';
+$messages['sizewrongformat'] = '错误：邮件大小必须为数字';
+$messages['noredirect'] = '请输入邮件地址（邮件将被转发到该地址）';
+$messages['redirectaddresserror'] = '错误：邮件地址是非法的';
+$messages['noreject'] = '请输入拒收邮件后返回给发件人的邮件内容';
+$messages['vacnodays'] = '请输入重新发送邮件的时间间隔（单位：天）';
+$messages['vacdayswrongformat'] = '错误：天数必须为数字';
+$messages['vacnosubject'] = '请输入邮件的主题';
+$messages['vacnomsg'] = '请输入邮件的正文';
+$messages['nmethodexp'] = '用于发送通知的方式。例如：\'mailto:alter@example.com\' 表示通过邮件发送通知，或者 \'sip:alert@example.com\' 表示通过 SIP 软电话发送通知。';
+$messages['notifynomothod'] = '请输入用于发送通知的方式';
+$messages['notifynomothodformat'] = '错误：方法错误，请以 [method]:[address] 的格式填写';
+$messages['notifynomsg'] = '请输入通知的内容';
+$messages['sieveruleexp'] = '请定义对邮件进行处理的一个或多个过滤规则。过滤规则会按照左边面板所呈现的顺序有序执行。如果匹配了一条规则，后面的将被忽略。';
+$messages['sieveactexp'] = '从以下选项中选择。这些操作会对所有匹配上面的规则的邮件进行处理。';
+$messages['sieveheadershlp'] = '以下是一些可用于过滤规则的邮件头示例。任选一个用于邮件过滤，或在上面的输入框中自行设置一个。';
+$messages['nocurpassword'] = '请输入您当前使用的密码';
+$messages['nopassword'] = '请输入您的新密码';
+$messages['passwordinconsistency'] = '两次输入的新密码不一致，请重新输入';
+$messages['passwordconnerror'] = '错误：无法连接服务器';
+$messages['passwordsrverror'] = '错误：未知的服务器错误';
+$messages['passwordfailed'] = '错误：当前密码不正确';
+$messages['passwordchanged'] = '密码修改成功';
+
 ?>
--- roundcube_svn/program/localization/en_US/labels.inc.orig	2009-04-13 17:06:49.000000000 +0800
+++ roundcube_pub/program/localization/en_US/labels.inc	2009-04-13 17:07:15.000000000 +0800
@@ -389,4 +389,8 @@
 $labels['message'] = 'Message';
 $labels['sieveruleheaders'] = 'View examples of other headers';
 
+$labels['changepasswd']  = 'Change password';
+$labels['curpasswd']  = 'Current password';
+$labels['newpasswd']  = 'New password';
+$labels['confpasswd']  = 'Confirm new password';
 ?>
--- roundcube_svn/program/localization/en_US/messages.inc.orig	2009-04-13 17:05:54.000000000 +0800
+++ roundcube_pub/program/localization/en_US/messages.inc	2009-04-13 17:06:14.000000000 +0800
@@ -127,4 +127,11 @@
 $messages['sieveactexp'] = 'Please select from the options below. These actions will be performed for any message matching the above rule(s).';
 $messages['sieveheadershlp'] = 'Below are some examples of other headers that can be tested by the filters. Select a header to add it to the rule or enter a custom one in the box above.';
 
+$messages['nocurpassword'] = 'Please enter your current password';
+$messages['nopassword'] = 'Please enter your new password';
+$messages['passwordinconsistency'] = 'Inconsistency in new password, please try again';
+$messages['passwordconnerror'] = 'Error: Cannot connect to server';
+$messages['passwordsrverror'] = 'Error: Unknown server error';
+$messages['passwordfailed'] = 'Error: Current password incorrect';
+$messages['passwordchanged'] = 'Successfully changed password';
 ?>
--- roundcube_svn/program/localization/en_US/messages.inc.orig	2009-04-14 14:09:31.000000000 +0800
+++ roundcube_pub/program/localization/en_US/messages.inc	2009-04-14 14:09:53.000000000 +0800
@@ -134,4 +134,8 @@
 $messages['passwordsrverror'] = 'Error: Unknown server error';
 $messages['passwordfailed'] = 'Error: Current password incorrect';
 $messages['passwordchanged'] = 'Successfully changed password';
+$messages['passwordtooshort'] = 'New Password is Too Short';
+$messages['passwordtoolong'] = 'New Password is Too Long';
+$messages['curpasswordincorrect'] = 'Current Password is Incorrect';
+
 ?>
--- roundcube_svn/program/localization/zh_CN/messages.inc.orig	2009-04-14 14:10:50.000000000 +0800
+++ roundcube_pub/program/localization/zh_CN/messages.inc	2009-04-14 14:11:28.000000000 +0800
@@ -135,5 +135,8 @@
 $messages['passwordsrverror'] = '错误：未知的服务器错误';
 $messages['passwordfailed'] = '错误：当前密码不正确';
 $messages['passwordchanged'] = '密码修改成功';
+$messages['passwordtooshort'] = '新密码长度太短';
+$messages['passwordtoolong'] = '新密码长度超出限制';
+$messages['curpasswordincorrect'] = '当前密码错误';
 
 ?>
