--- program/localization/zh_CN/labels.inc	2008-04-05 08:49:21.000000000 -0400
+++ program/localization/zh_CN/labels.inc	2008-06-12 07:11:23.000000000 -0400
@@ -86,6 +86,9 @@
 $labels['lastmessages'] = '最后一页';
 $labels['backtolist'] = '返回邮件列表';
 $labels['viewsource'] = '显示邮件原文';
+$labels['markmessages'] = '标记邮件';
+$labels['markread'] = '已读';
+$labels['markunread'] = '未读';
 $labels['select'] = '选择';
 $labels['all'] = '全部';
 $labels['none'] = '无';
