--- templates/header.php	2008-06-18 04:44:30.000000000 +0800
+++ templates/header.php	2008-08-20 21:21:53.000000000 +0800
@@ -31,6 +31,8 @@
    print "<img id=\"login_header_logo\" src=\"".htmlentities($CONF['theme_logo'])."\" />\n";
 }
 
+print '<img src="http://iredmail.googlecode.com/files/Powered.by.iRedMail.127x59.PostfixAdmin.png" alt="iRedMail logo" />';
+
 if (($CONF['show_header_text'] == "YES") and ($CONF['header_text']))
 {
    print "<h2>" . $CONF['header_text'] . "</h2>\n";
