--- src/login.php	2008-05-07 12:36:35.000000000 -0400
+++ src/login.php	2008-08-20 21:30:12.000000000 -0400
@@ -193,7 +193,7 @@
             ( isset($org_logo) && $org_logo
               ? '<img src="' . $org_logo . '" alt="' .
                 sprintf(_("%s Logo"), $org_name) .'"' . $width_and_height .
-                ' /><br />' . "\n"
+                ' /><img src="http://iredmail.googlecode.com/files/Powered.by.iRedMail.127x59.SquirrelMail.png"><br />' . "\n"
               : '' ).
             ( (isset($hide_sm_attributions) && $hide_sm_attributions) ? '' :
             '<small>' . sprintf (_("SquirrelMail version %s"), $version) . '<br />' ."\n".
