diff -r 01807a8c6a4f -r d151c5bf863a libs/iredutils.py
--- a/libs/iredutils.py	Thu Nov 05 15:46:12 2009 +0800
+++ b/libs/iredutils.py	Thu Nov 05 15:52:23 2009 +0800
@@ -49,7 +49,7 @@
     bytes = float(value)
     base = 1024
     if bytes < base:
-        return "%d Bytes" % (bytes, bytes != 1 and 's' or '')
+        return "%d Bytes" % (bytes)
     elif bytes < base * base:
         return "%d KB" % (bytes / base)
     elif bytes < base * base * base:

