--- program/include/iniset.php	2008-12-30 22:33:28.000000000 +0800
+++ program/include/iniset.php	2009-01-10 17:27:41.000000000 +0800
@@ -50,11 +50,15 @@
   die('Fatal error: ini_set/set_include_path does not work.');
 }
 
+ini_set('error_reporting', E_ALL&~E_NOTICE);
+if  (isset($_SERVER['HTTPS'])) {
+    ini_set('session.cookie_secure', ($_SERVER['HTTPS'] && ($_SERVER['HTTPS'] != 'off'))?1:0);
+} else {
+    ini_set('session.cookie_secure', 0);
+}
 ini_set('session.name', 'roundcube_sessid');
 ini_set('session.use_cookies', 1);
 ini_set('session.only_use_cookies', 1);
-ini_set('session.cookie_secure', ($_SERVER['HTTPS'] && ($_SERVER['HTTPS'] != 'off')));
-ini_set('error_reporting', E_ALL&~E_NOTICE);
 set_magic_quotes_runtime(0);
 
 // increase maximum execution time for php scripts
