--- turba/lib/Driver/ldap.php	2009-05-01 06:12:37.000000000 -0400
+++ turba/lib/Driver/ldap.php	2009-06-24 23:49:20.000000000 -0400
@@ -162,7 +162,7 @@
 
         /* Add source-wide filters, which are _always_ AND-ed. */
         if (!empty($this->_params['filter'])) {
-            $filter = '(&' . '(' . $this->_params['filter'] . ')' . $filter . ')';
+            $filter = '(&' . $this->_params['filter'] . $filter . ')';
         }
 
         /* Four11 (at least) doesn't seem to return 'cn' if you don't
