diff -ruN roundcubemail-0.2-stable.orig/index.php roundcubemail-0.2-stable/index.php
--- roundcubemail-0.2-stable.orig/index.php	2008-12-30 22:33:28.000000000 +0800
+++ roundcubemail-0.2-stable/index.php	2009-01-04 23:56:24.000000000 +0800
@@ -193,6 +193,7 @@
     'subscribe'     => 'manage_folders.inc',
     'unsubscribe'   => 'manage_folders.inc',
     'add-identity'  => 'edit_identity.inc',
+    'save-passwd'   => 'passwd.inc',
   )
 );
 
diff -ruN roundcubemail-0.2-stable.orig/program/js/app.js roundcubemail-0.2-stable/program/js/app.js
--- roundcubemail-0.2-stable.orig/program/js/app.js	2008-12-30 23:25:56.000000000 +0800
+++ roundcubemail-0.2-stable/program/js/app.js	2009-01-05 00:01:36.000000000 +0800
@@ -235,7 +235,7 @@
 this.enable_command("list",true);
 break;
 case "settings":
-this.enable_command("preferences","identities","save","folders",true);
+this.enable_command("preferences","identities","save","folders","passwd",true);
 if(this.env.action=="identities"||this.env.action=="edit-identity"||this.env.action=="add-identity"){
 this.enable_command("add",this.env.identities_level<2);
 this.enable_command("delete","edit",true);
@@ -246,6 +246,19 @@
 if(this.env.action=="folders"){
 this.enable_command("subscribe","unsubscribe","create-folder","rename-folder","delete-folder",true);
 }
+if (this.env.action=='passwd' || this.env.action=='save-passwd')
+{
+var input_curpasswd=rcube_find_object('_curpasswd');
+var input_newpasswd=rcube_find_object('_newpasswd');
+var input_confpasswd=rcube_find_object('_confpasswd');
+if(input_curpasswd&&input_curpasswd.value==''){
+input_curpasswd.focus();
+}
+else if (input_confpasswd){
+input_confpasswd.focus();
+}
+this.enable_command('save-passwd', true);
+}
 if(this.gui_objects.identitieslist){
 this.identity_list=new rcube_list_widget(this.gui_objects.identitieslist,{multiselect:false,draggable:false,keyboard:false});
 this.identity_list.addEventListener("select",function(o){
@@ -892,6 +905,33 @@
 case "delete-folder":
 this.delete_folder(_40);
 break;
+case 'passwd':
+this.goto_url('passwd');
+break;
+
+case 'save-passwd':
+var input_curpasswd = rcube_find_object('_curpasswd');
+var input_newpasswd = rcube_find_object('_newpasswd');
+var input_confpasswd = rcube_find_object('_confpasswd');
+if ((input_newpasswd && input_newpasswd.value=='') && (input_confpasswd && input_confpasswd.value=='') || (input_curpasswd && input_curpasswd.value==''))
+{
+alert(this.get_label('nopassword'));
+input_curpasswd.value='';
+input_newpasswd.value='';
+input_confpasswd.value='';
+input_curpasswd.focus();
+}
+else if ((input_newpasswd && input_confpasswd) && ( input_newpasswd.value != input_confpasswd.value))
+{
+alert(this.get_label('passwordinconsistency'));
+input_newpasswd.value='';
+input_confpasswd.value='';
+input_newpasswd.focus();
+}
+else{
+this.gui_objects.editform.submit();
+}
+break;
 }
 return obj?false:true;
 };
diff -ruN roundcubemail-0.2-stable.orig/program/localization/de_DE/labels.inc roundcubemail-0.2-stable/program/localization/de_DE/labels.inc
--- roundcubemail-0.2-stable.orig/program/localization/de_DE/labels.inc	2008-12-07 01:59:26.000000000 +0800
+++ roundcubemail-0.2-stable/program/localization/de_DE/labels.inc	2009-01-05 00:02:36.000000000 +0800
@@ -264,4 +264,10 @@
 $labels['MB'] = 'MB';
 $labels['GB'] = 'GB';
 
+// Plesk Change Password
+$labels['changepasswd']  = 'Passwort ändern';
+$labels['curpasswd']  = 'Altes Passwort';
+$labels['newpasswd']  = 'Neues Passwort';
+$labels['confpasswd']  = 'Neues Passwort wiederholen';
+
 ?>
diff -ruN roundcubemail-0.2-stable.orig/program/localization/de_DE/messages.inc roundcubemail-0.2-stable/program/localization/de_DE/messages.inc
--- roundcubemail-0.2-stable.orig/program/localization/de_DE/messages.inc	2008-11-18 02:52:43.000000000 +0800
+++ roundcubemail-0.2-stable/program/localization/de_DE/messages.inc	2009-01-05 00:02:20.000000000 +0800
@@ -95,4 +95,11 @@
 $messages['nofromaddress'] = 'Fehlende E-Mail-Adresse in ausgewählter Identität';
 $messages['editorwarning'] = 'Beim Wechseln in den Texteditor gehen alle Textformatierungen verloren. Möchten Sie fortfahren?';
 
+//Plesk Password Change
+$messages['nopassword'] = "Bitte das neue Passwort angeben.";
+$messages['passwordinconsistency'] = "Das neue Passwort stimmt nicht überein";
+$messages['passwordtooshort'] = 'Das neue Passwort ist zu kurz';
+$messages['passwordtoolong'] = 'Das neue Passwort ist zu lang';
+$messages['curpasswordincorrect'] = 'Das  alte Passwort ist falsch';
+
 ?>
diff -ruN roundcubemail-0.2-stable.orig/program/localization/en_US/labels.inc roundcubemail-0.2-stable/program/localization/en_US/labels.inc
--- roundcubemail-0.2-stable.orig/program/localization/en_US/labels.inc	2008-12-07 01:59:26.000000000 +0800
+++ roundcubemail-0.2-stable/program/localization/en_US/labels.inc	2009-01-05 00:02:08.000000000 +0800
@@ -323,4 +323,10 @@
 $labels['MB'] = 'MB';
 $labels['GB'] = 'GB';
 
+// Plesk Change Password
+$labels['changepasswd']  = 'Change Password';
+$labels['curpasswd']  = 'Current Password';
+$labels['newpasswd']  = 'New Password';
+$labels['confpasswd']  = 'Confirm New Password';
+
 ?>
diff -ruN roundcubemail-0.2-stable.orig/program/localization/en_US/messages.inc roundcubemail-0.2-stable/program/localization/en_US/messages.inc
--- roundcubemail-0.2-stable.orig/program/localization/en_US/messages.inc	2008-11-04 02:09:35.000000000 +0800
+++ roundcubemail-0.2-stable/program/localization/en_US/messages.inc	2009-01-05 00:01:54.000000000 +0800
@@ -94,5 +94,10 @@
 $messages['opnotpermitted'] = 'Operation not permitted!';
 $messages['nofromaddress'] = 'Missing e-mail address in selected identity';
 $messages['editorwarning'] = 'Switching to the plain text editor will cause all text formatting to be lost. Do you wish to continue?';
-
+//Plesk Password Change
+$messages['nopassword'] = "Please input new password.";
+$messages['passwordinconsistency'] = "New Passwords Do Not Match";
+$messages['passwordtooshort'] = 'New Password Too Short';
+$messages['passwordtoolong'] = 'New Password Too Long';
+$messages['curpasswordincorrect'] = 'Current Password Incorrect';
 ?>
diff -ruN roundcubemail-0.2-stable.orig/program/steps/settings/func.inc roundcubemail-0.2-stable/program/steps/settings/func.inc
--- roundcubemail-0.2-stable.orig/program/steps/settings/func.inc	2008-12-17 03:05:50.000000000 +0800
+++ roundcubemail-0.2-stable/program/steps/settings/func.inc	2009-01-05 00:02:58.000000000 +0800
@@ -491,7 +491,8 @@
 $OUTPUT->add_handlers(array(
   'userprefs' => 'rcmail_user_prefs_form',
   'identitieslist' => 'rcmail_identities_list',
-  'itentitieslist' => 'rcmail_identities_list'  // keep this for backward compatibility
+  'itentitieslist' => 'rcmail_identities_list',  // keep this for backward compatibility
+  'userpasswd' => 'rcmail_passwd_form'
 ));
 
 
diff -ruN roundcubemail-0.2-stable.orig/program/steps/settings/passwd.inc roundcubemail-0.2-stable/program/steps/settings/passwd.inc
--- roundcubemail-0.2-stable.orig/program/steps/settings/passwd.inc	1970-01-01 08:00:00.000000000 +0800
+++ roundcubemail-0.2-stable/program/steps/settings/passwd.inc	2009-02-03 22:54:12.000000000 +0800
@@ -0,0 +1,157 @@
+<?php
+
+/*
+ -----------------------------------------------------------------------
+ | program/steps/settings/passwd.inc                                   |
+ |                                                                     |
+ | This file is part of the RoundCube Webmail client                   |
+ | Copyright (C) 2007, RoundCube Dev. - Switzerland                    |
+ |                                                                     |
+ | Passwd_Driver_poppassd Class (C) 2000-2007 Horde Project            |
+ |                                                                     |
+ | Licensed under the GNU GPL                                          |
+ |                                                                     |
+ | PURPOSE:                                                            |
+ |   Change IMAP user password for Plesk and other poppassd servers    |
+ |                                                                     |
+ | REQUIREMENTS:                                                       |
+ |   PHP, and Plesk or other poppassd server                           |
+ |                                                                     |
+ | Passwd_Driver_poppassd adopted from Horde Webmail                   |
+ |                                                                     |
+ -----------------------------------------------------------------------
+ | cPanel Version Author: Danny Herran <danny@ingeniarte.com>          |
+ | Passwd_Driver_poppassd Author: Eric Jon Rostetter                   |
+ |                                <eric.rostetter@physics.utexas.edu>  |
+ | Modifications Made By: Mark Romero <mromero@simplewebworks.com>     |
+ | Modifications Made By: Samuel Barabas <samuel@owee.de>              |
+ -----------------------------------------------------------------------
+
+ $Id: passwd.inc,v 0.2 2008/04/08 23:19:03 roundcube Exp $
+
+*/
+
+// The MIN length of a password for your serveras defined in /etc/login.defs
+// -1 for no min/max
+define("POPPASSD_MIN_LENGTH", 5); 
+define("POPPASSD_MAX_LENGTH", -1);
+
+function rcmail_save_passwd($curpasswd, $newpasswd){
+  global $CONFIG, $_SESSION, $OUTPUT, $RCMAIL;
+  
+  //$driver = new Passwd_Driver_poppassd(array('host'=>'localhost', 'port'=> '106'));
+  //$return = $driver->changePassword($_SESSION['username'], $curpasswd, $newpasswd);
+  $return = update_pass($_SESSION['username'], $curpasswd, $newpasswd);  
+
+  //If we didn't encountry a problem ($return[0] == true)
+  if ($return[0]) {
+	  //We update the password for the $this->encrypt_passwd($pass)current session
+	  $_SESSION['password'] = $RCMAIL->encrypt_passwd($newpasswd);
+  }
+  return $return;
+} 
+
+function update_pass($username, $old_password, $new_password) {
+  $ldap_host = "localhost";
+  list($user,$domain) = split("@", $username);
+  $dc_array = split('\\.', $domain);
+  foreach ($dc_array as $value)
+    $dc.="dc=".$value.",";
+  $dc=substr($dc, 0, -1);
+
+  $ldap_user = "mail=$username,domainName=$domain,o=domains,$dc";
+  $ldap_pass = $old_password;
+  if (!($connect = ldap_connect($ldap_host)))
+    die("Cannot connect to LDAP server");
+  ldap_set_option($connect, LDAP_OPT_PROTOCOL_VERSION, 3);
+  ldap_set_option($connect, LDAP_OPT_REFERRALS, 0);
+
+  if (!ldap_bind($connect, $ldap_user, $ldap_pass))
+    return array(false, "Incorrect Password");
+  $info['userPassword']="{MD5}".base64_encode(pack("H*",md5($new_password)));
+  if (ldap_modify($connect, $ldap_user, $info)) {
+    ldap_close($connect);
+    return array(true, "Success");
+  }
+  ldap_close($connect);
+  die("Cannot update new password");
+}
+
+function rcmail_passwd_form($attrib){
+  global $CONFIG, $OUTPUT;
+														    
+  list($form_start, $form_end) = get_form_tags($attrib, 'save-passwd');
+  unset($attrib['form']);
+														        
+  if (!$attrib['id'])
+    $attrib['id'] = 'rcmSavepassword';
+
+  // allow the following attributes to be added to the <table> tag
+  $attrib_str = create_attrib_string($attrib, array('style', 'class', 'id', 'cellpadding', 'cellspacing', 'border', 'summary'));
+
+  // return the complete edit form as table
+  $out = "$form_start<table" . $attrib_str . ">\n\n";
+
+  $a_show_cols = array('curpasswd'   => array('type' => 'text'), 'newpasswd'   => array('type' => 'text'), 'confpasswd'   => array('type' => 'text'));
+
+  // show current password field
+  $field_id = 'curpasswd';
+  $input_curpasswd = new html_passwordfield(array('name' => '_curpasswd', 'id' => $field_id, 'size' => 30));
+
+  $out .= sprintf("<tr><td class=\"title\"><label for=\"%s\">%s</label></td><td>%s</td></tr>\n", $field_id, rep_specialchars_output(rcube_label('curpasswd')), $input_curpasswd->show($CONFIG['curpasswd']));
+
+  // show new password selection
+  $field_id = 'newpasswd';
+  $input_newpasswd = new html_passwordfield(array('name' => '_newpasswd', 'id' => $field_id, 'size' => 30));
+
+  $out .= sprintf("<tr><td class=\"title\"><label for=\"%s\">%s</label></td><td>%s</td></tr>\n", $field_id, rep_specialchars_output(rcube_label('newpasswd')), $input_newpasswd->show($CONFIG['newpasswd']));
+
+  // show confirm password selection
+  $field_id = 'confpasswd';
+  $input_confpasswd = new html_passwordfield(array('name' => '_confpasswd', 'id' => $field_id, 'size' => 30));
+
+  $out .= sprintf("<tr><td class=\"title\"><label for=\"%s\">%s</label></td><td>%s</td></tr>\n", $field_id, rep_specialchars_output(rcube_label('confpasswd')), $input_confpasswd->show($CONFIG['confpasswd']));
+
+  $out .= "\n</table>$form_end";
+
+  return $out;  
+}
+
+// Handle user request
+if ($RCMAIL->action=='save-passwd'){
+
+  if (!isset($_POST['_newpasswd']) || !isset($_POST['_curpasswd']))
+    $OUTPUT->show_message('errorsaving', 'error');
+  elseif(POPPASSD_MIN_LENGTH != -1 && (strlen($_POST['_newpasswd']) < POPPASSD_MIN_LENGTH)) 
+    $OUTPUT->show_message('passwordtooshort', 'error');
+  elseif(POPPASSD_MIN_LENGTH != -1 && (strlen($_POST['_newpasswd']) < POPPASSD_MIN_LENGTH)) 
+    $OUTPUT->show_message('passwordtooshort', 'error');
+  elseif($_POST['_newpasswd'] != $_POST['_confpasswd'])
+    $OUTPUT->show_message('passwordinconsistency', 'error');
+  else
+  {
+    $return = rcmail_save_passwd(get_input_value('_curpasswd', RCUBE_INPUT_POST), get_input_value('_newpasswd', RCUBE_INPUT_POST));
+    if($return[0])
+      $OUTPUT->show_message('successfullysaved', 'confirmation');
+    else
+    {
+      if($return[1] == "Incorrect Password")
+        $OUTPUT->show_message('curpasswordincorrect', 'error');
+      else
+        $OUTPUT->show_message('errorsaving', 'error');
+    }
+  }
+
+  // overwrite action variable
+  rcmail_overwrite_action('passwd');
+
+}
+
+// add some labels to client
+$OUTPUT->add_label('nopassword');
+$OUTPUT->add_label('passwordtooshort');
+$OUTPUT->add_label('passwordtoolong');
+$OUTPUT->add_label('passwordinconsistency');
+$OUTPUT->add_label('curpasswordincorrect');
+$OUTPUT->send('passwd');
+?>
diff -ruN roundcubemail-0.2-stable.orig/skins/default/includes/settingstabs.html roundcubemail-0.2-stable/skins/default/includes/settingstabs.html
--- roundcubemail-0.2-stable.orig/skins/default/includes/settingstabs.html	2008-05-20 21:33:47.000000000 +0800
+++ roundcubemail-0.2-stable/skins/default/includes/settingstabs.html	2009-01-05 00:00:48.000000000 +0800
@@ -2,4 +2,5 @@
 <span id="settingstabdefault" class="tablink"><roundcube:button command="preferences" type="link" label="preferences" title="editpreferences" /></span>
 <span id="settingstabfolders" class="tablink"><roundcube:button command="folders" type="link" label="folders" title="managefolders" class="tablink" /></span>
 <span id="settingstabidentities" class="tablink"><roundcube:button command="identities" type="link" label="identities" title="manageidentities" class="tablink" /></span>
+<span id="settingstabpasswd" class="tablink"><roundcube:button command="passwd" type="link" label="password" title="changepasswd" class="tablink" /></span>
 </div>
diff -ruN roundcubemail-0.2-stable.orig/skins/default/templates/passwd.html roundcubemail-0.2-stable/skins/default/templates/passwd.html
--- roundcubemail-0.2-stable.orig/skins/default/templates/passwd.html	1970-01-01 08:00:00.000000000 +0800
+++ roundcubemail-0.2-stable/skins/default/templates/passwd.html	2009-01-05 00:01:04.000000000 +0800
@@ -0,0 +1,27 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml">
+	<head>
+		<title><roundcube:object name="pagetitle" /></title>
+		<roundcube:include file="/includes/links.html" />
+		<link rel="stylesheet" type="text/css" href="/settings.css" />
+	</head>
+	<body>
+
+		<roundcube:include file="/includes/taskbar.html" />
+		<roundcube:include file="/includes/header.html" />
+		<roundcube:include file="/includes/settingstabs.html" />
+
+		<div id="userprefs-box">
+			<div id="userprefs-title"><roundcube:label name="changepasswd" /></div>
+
+			<div style="padding:15px">
+				<roundcube:object name="userpasswd">
+
+				<p><br /><roundcube:button command="save-passwd" type="input" class="button" label="save" /></p>
+			</div>
+		</div>
+
+		<roundcube:include file="/includes/settingscripts.html" />
+
+	</body>
+</html>
\ No newline at end of file
