--- libs/Ext/Storage/Maildir.pm 2008-09-20 12:46:27.000000000 +0800
+++ /root/Maildir.pm.new    2008-11-29 10:16:59.000000000 +0800
@@ -12,6 +12,10 @@
 package Ext::Storage::Maildir;
 use strict;
 
+# Used to recursively create user maildir. Patched by iRedMail project:
+# http://code.google.com/p/iredmail/
+use File::Path;
+
 # Maildir++ specification
 # see http://www.inter7.com/courierimap/README.maildirquota.html
 #
@@ -84,6 +88,13 @@
     $CFG{mode} = $mode ? $mode : "O_RW";
     $CFG{warnlv} = $warnlv ? $warnlv : "0.9"; # 90% default
 
+    # Auto create user maildir if not exist. Patched by iRedMail project:
+    # http://code.google.com/p/iredmail/
+    if ( ! -d $CFG{path}) {
+        mkpath($CFG{path}, 0, 0700) or die "Can't create user maildir $CFG{path}, $!\n";
+    }
+
     chdir(untaint($CFG{path})) or die "Can't chdir to $CFG{path}, $!\n";
 
     mkdir "cur", 0700 if(!-d 'cur');
